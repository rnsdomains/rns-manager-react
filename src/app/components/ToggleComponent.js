/* eslint-disable react/self-closing-comp */
import React, { useState } from 'react';
import propTypes from 'prop-types';

const ToggleComponent = (props) => {
  const {
    labelLeft,
    labelRight,
    onChange,
    initialValue,
  } = props;

  const [value, setValue] = useState(initialValue);

  const handleChange = (changeTo) => {
    setValue(changeTo);
    onChange(changeTo);
  };

  return (
    <div className="toggleSwitch">
      <button
        className={!value ? 'left active' : 'left'}
        onClick={() => handleChange(false)}
        type="button"
      >
        {labelLeft}
      </button>
      <label className="switch" htmlFor="toggleSwitch">
        <input
          type="checkbox"
          onChange={() => handleChange(!value)}
          id="toggleSwitch"
          checked={value}
        />
        <span className="slider"></span>
      </label>
      <button
        className={value ? 'right active' : 'right'}
        onClick={() => handleChange(true)}
        type="button"
      >
        {labelRight}
      </button>
    </div>
  );
};

ToggleComponent.defaultProps = {
  labelLeft: 'leftLabel',
  labelRight: 'rightLabel',
  initialValue: false,
};

ToggleComponent.propTypes = {
  labelLeft: propTypes.string,
  labelRight: propTypes.string,
  onChange: propTypes.func.isRequired,
  initialValue: propTypes.bool,
};

export default ToggleComponent;
