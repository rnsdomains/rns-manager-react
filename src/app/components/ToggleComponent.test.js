import React from 'react';
import { mount } from 'enzyme';
import ToggleComponent from './ToggleComponent';


describe('toggleComponent', () => {
  const handleChange = jest.fn();
  it('defaults matches snapshot', () => {
    const component = mount(<ToggleComponent onChange={handleChange} />);
    expect(component.find('input[type="checkbox"]').props().value).toBeFalsy();

    const activeButton = component.find('button.active');
    expect(activeButton.text()).toEqual('leftLabel');

    expect(component).toMatchSnapshot();
  });

  it('toggleSwitch active matches snapshot', () => {
    const component = mount(<ToggleComponent onChange={handleChange} initialValue />);
    expect(component.find('input[type="checkbox"]').props().value).toBeFalsy();

    const activeButton = component.find('button.active');
    expect(activeButton.text()).toEqual('rightLabel');

    expect(component).toMatchSnapshot();
  });

  it('changes value when switch is toggled', () => {
    const localHandleChange = jest.fn();
    const component = mount(<ToggleComponent onChange={localHandleChange} />);
    const checkBox = () => component.find('input[type="checkbox"]');

    expect(checkBox().props().checked).toBeFalsy();
    checkBox().simulate('change', { target: { checked: false } });
    expect(checkBox().props().checked).toBeTruthy();
    expect(localHandleChange.mock.calls.length).toBe(1);

    expect(component.find('button.active').text()).toBe('rightLabel');
  });

  it('sets the text from the parent', () => {
    const component = mount(
      <ToggleComponent labelLeft="LEFT" labelRight="RIGHT" onChange={handleChange} />,
    );

    expect(component.find('button.left').text()).toEqual('LEFT');
    expect(component.find('button.right').text()).toEqual('RIGHT');
  });

  it('keeps active when button is clicked', () => {
    const localHandleChange = jest.fn();
    const component = mount(<ToggleComponent onChange={localHandleChange} />);

    component.find('button.left').simulate('click');
    expect(localHandleChange.mock.calls.length).toBe(1);
    expect(component.find('button.left').props().className).toEqual('left active');
    expect(component.find('button.right').props().className).toEqual('right');

    // click button twice
    component.find('button.right').simulate('click').simulate('click');
    expect(localHandleChange.mock.calls.length).toBe(3);
    expect(component.find('button.right').props().className).toEqual('right active');
  });
});
