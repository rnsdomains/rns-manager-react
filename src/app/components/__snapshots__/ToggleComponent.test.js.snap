// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toggleComponent defaults matches snapshot 1`] = `
<ToggleComponent
  initialValue={false}
  labelLeft="leftLabel"
  labelRight="rightLabel"
  onChange={[MockFunction]}
>
  <div
    className="toggleSwitch"
  >
    <button
      className="left active"
      onClick={[Function]}
      type="button"
    >
      leftLabel
    </button>
    <label
      className="switch"
      htmlFor="toggleSwitch"
    >
      <input
        checked={false}
        id="toggleSwitch"
        onChange={[Function]}
        type="checkbox"
      />
      <span
        className="slider"
      />
    </label>
    <button
      className="right"
      onClick={[Function]}
      type="button"
    >
      rightLabel
    </button>
  </div>
</ToggleComponent>
`;

exports[`toggleComponent toggleSwitch active matches snapshot 1`] = `
<ToggleComponent
  initialValue={true}
  labelLeft="leftLabel"
  labelRight="rightLabel"
  onChange={[MockFunction]}
>
  <div
    className="toggleSwitch"
  >
    <button
      className="left"
      onClick={[Function]}
      type="button"
    >
      leftLabel
    </button>
    <label
      className="switch"
      htmlFor="toggleSwitch"
    >
      <input
        checked={true}
        id="toggleSwitch"
        onChange={[Function]}
        type="checkbox"
      />
      <span
        className="slider"
      />
    </label>
    <button
      className="right active"
      onClick={[Function]}
      type="button"
    >
      rightLabel
    </button>
  </div>
</ToggleComponent>
`;
